<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Corrected the Coverage of Credible Sets from Bayesian Genetic Fine Mapping • corrcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Corrected the Coverage of Credible Sets from Bayesian Genetic Fine Mapping">
<meta property="og:description" content='The package uses a computationally efficient method to find the
    "Corrected Coverage Estimate" of a credible set of putative
    causal variants from Bayesian genetic fine-mapping. The package can 
    also be used to obtain a "Corrected Credible Set" if required; 
    that is, the smallest set of variants required such that the
    corrected coverage estimate of the resultant credible set is  
    within some user defined accuracy of the desired coverage. '>
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">corrcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/New-Credible-Set.html">Corrected Credible Set</a>
    </li>
    <li>
      <a href="articles/Useful-Info.html">Useful Info</a>
    </li>
    <li>
      <a href="articles/corrected-coverage.html">Correcting the Coverage of Credible Sets</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="corrcoverage" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#corrcoverage" class="anchor"></a>corrcoverage <img src="reference/figures/logo.png" align="right">
</h1></div>

<p>Webpage: <a href="https://annahutch.github.io/corrcoverage/" class="uri">https://annahutch.github.io/corrcoverage/</a></p>
<p>The <code>corrcoverage</code> R package uses a computationally efficient algorithm to find accurate coverage estimates of the causal variant in credible sets of prioritised variants from single causal variant genetic fine mapping (<a href="https://www.ncbi.nlm.nih.gov/pubmed/23104008">Maller et al., 2012</a>, <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/gepi.20359">Wakefield, 2009</a>.).</p>
<p>The package only requires GWAS summary statistics and can be used to:</p>
<ul>
<li>Perform single causal variant Bayesian fine-mapping</li>
<li>Estimate the true genetic effect at the causal variant (see <code>est_mu</code> function)</li>
<li>Obtain an accurate coverage estimate that the causal variant is contained within a credible set, the “corrected coverage estimate” (see <code>corrcov</code> function and “Corrected Coverage”" vignette)</li>
<li>Find a new “corrected” credible set with the desired coverage of the true causal variant (see <code>corrected_cs</code> function and “Corrected Credible Set”" vignette)</li>
</ul>
<p>We’ve strived to make our R package as easy to use as possible. Please see the flowchart below to decide which function is best to solve your problem. The interactive version (click-to-functions) is <a href="https://annahutch.github.io/PhD/package_flowchart.html">available here</a>.</p>
<p><img src="https://annahutch.github.io/PhD/package_flowchart.svg"></p>
<hr>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of <code>corrcoverage</code> from <a href="https://github.com/">github</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>) <span class="co"># if not already installed</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">"annahutch/corrcoverage"</span>)</a></code></pre></div>
<hr>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>For examples, please see the relevant vignettes.</p>
<ol>
<li><p>The “Corrected Coverage” vignette <a href="https://annahutch.github.io/corrcoverage/articles/Corrected-Coverage.html">here</a> should be read first. This shows readers how to use the <code>corrcoverage</code> R package to get an accurate coverage estimate of the causal variant in a credible set.</p></li>
<li><p>The “Corrected Credible Set” vignette <a href="https://annahutch.github.io/corrcoverage/articles/New-Credible-Set.html">here</a> follows on from the “Corrected Coverage”" vignette and shows readers how the <code>corrcoverage</code> R package can be used to obtain a new credible set with the desired coverage of the causal variant.</p></li>
<li><p>The ‘Useful Info’ vignette <a href="https://annahutch.github.io/corrcoverage/articles/Useful-Info.html">here</a> provides supplementary information about the usage of the package, including information about other useful functions.</p></li>
</ol>
<hr>
<p>In brief, the correction method involves simulating many credible sets from “the same system as the original” and calculating what proportion of these contain the true causal variant. Since the true causal variant is unknown, each variant is considered as causal in turn and the proportions are normalised by that variant’s posterior probability of causality.</p>
<hr>
</div>
<div id="abstract" class="section level2">
<h2 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h2>
<p>Genome Wide Association Studies (GWAS) have successfully identified thousands of loci associated with human diseases. Bayesian genetic fine-mapping studies aim to identify the specific causal variants within GWAS loci responsible for the disease, reporting credible sets of plausible causal variants, which are interpreted as containing the causal variant with some ‘coverage probability’.</p>
<p>Here, we investigate the coverage probabilities of credible sets through simulations and find them to be systematically biased. We present a method to re-estimate the coverage of credible sets using rapid simulations based on the observed, or estimated, SNP correlation structure, we call this re-estimation the “corrected coverage estimate”. This is extended to find “corrected credible sets”, which are the smallest set of variants such that the corrected coverage probability is accurate.</p>
<p>We show that our method can be used to improve the resolution of fine-mapping studies using empirical type 1 diabetes data. Specifically, we found that in 29 out of 39 associated genomic regions, our method could be used to reduce the number of potentially causal variants to consider for follow-up – leading to more powerful and reliable follow up studies.</p>
<p>Crucially, our correction method requires only genetic association summary test statistics and remains accurate when SNP correlations are estimated from a large reference panel. Using our method to improve the resolution of fine-mapping experiments will enable more efficient expenditure of resources in the follow-up process of annotating the variants in the credible set to determine the implicated genes and pathways.</p>
<hr>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Report a bug at <br><a href="https://github.com/annahutch/corrcoverage/issues">https://​github.com/​annahutch/​corrcoverage/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Anna Hutchinson <br><small class="roles"> Author, maintainer </small>  </li>
<li>Chris Wallace <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://codecov.io/github/annahutch/corrcoverage?branch=master"><img src="https://codecov.io/gh/annahutch/corrcoverage/branch/master/graph/badge.svg" alt="Coverage status"></a></li>
<li><a href="https://travis-ci.org/annahutch/corrcoverage"><img src="https://travis-ci.org/annahutch/corrcoverage.svg?branch=master" alt="Build Status"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
