---
title: "Correcting the Coverage of Credible Sets"
author: "Anna Hutchinson"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, set.seed(2), include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(width = 3000)
```

---

This guide will show readers how to use the `corrcoverage` package for fine-mapping analyses, including obtaining an accurate coverage estimate for the causal variant in a credible set. This package is specific to credible sets obtained using the Bayesian approach to fine-mapping, described by Maller et al. [here](https://www.ncbi.nlm.nih.gov/pubmed/23104008).

The `corrcoverage` package uses GWAS summary statistics to provides an accurate estimate of the probability that the true causal variant is contained within a credible set derived for a given threshold. 

---

Load the library.

```{r}
set.seed(2)
library(corrcoverage)
```

---

## Simulate GWAS summary statistics

---

**This package is intended for use on real summary statistics obtained from GWAS, such as observed $Z$ scores. For the purpose of this vignette, we will simulate artificial haplotypes and GWAS data using the [simGWAS](https://github.com/chr1swallace/simGWAS) package. Please refer to the walkthrough guide [here](https://chr1swallace.github.io/simGWAS/articles/intro.html) from which the following is taken from.** 

```{r}
library(simGWAS)
```

```{r}
# Simulate reference haplotypes
nsnps <- 200
nhaps <- 1000
lag <- 5 # genotypes are correlated between neighbouring variants
maf <- runif(nsnps+lag,0.05,0.5) # common SNPs
laghaps <- do.call("cbind", lapply(maf, function(f) rbinom(nhaps,1,f)))
haps <- laghaps[,1:nsnps]
for(j in 1:lag) 
    haps <- haps + laghaps[,(1:nsnps)+j]
haps <- round(haps/matrix(apply(haps,2,max),nhaps,nsnps,byrow=TRUE))
snps <- colnames(haps) <- paste0("s",1:nsnps)
freq <- as.data.frame(haps+1)
freq$Probability <- 1/nrow(freq)
sum(freq$Probability)
MAF <- colMeans(freq[,snps]-1)
```

Specify the causal variant and it's effect on disease, as odds ratio. One causal variant is chosen since the Bayesian approach to fine-mapping relies on the assumption of one causal variant per region. We choose a CV from the column names in `freq` and a value for the odds ratio.

```{r}
CV <- sample(snps[which(colMeans(haps)>0.1)],1)
iCV <- sub("s", "", CV) # index of cv
g1 <- 1.1
```

We then simulate marginal $Z$ scores. Here, we consider a GWAS with 5000 cases and 5000 controls. 

```{r}
z0 <- simulated_z_score(N0=5000, # number of controls
              N1=5000, # number of cases
              snps=snps, # column names in freq of SNPs for which Z scores should be generated
              W=CV, # causal variants, subset of snps
              gamma.W=log(g1), # log odds ratios
              freq=freq # reference haplotypes
              )
```

----

## Obtain credible set

----

We now use the `credsets` function in the `corrcoverage` package to obtain a credible set using the Bayesian approach for fine-mapping. In brief, the function follows the standard steps from Maller's approach:

1. Order variants in descending order of posterior probabilities. 
2. Cumulatively sum these variants until the specified threshold is exceeded.
3. Group these variants and form the credible set.

The function requires posterior probabilities as input and so we use the `ppfunc` function to convert the marginal $Z$ scores to posterior probabilties. This function requires a value for $V$, the variance of the estimate effect size, which can be easily calculated using the `Var.data.cc` function in the `coloc` package. The prior standard deviation of the effect size, $W$, is an optional parameter with a default value of 0.2.

```{r warning=FALSE, error=FALSE, fig.width=4}
N0 <- 5000 # number of controls
N1 <- 5000 # number of cases

var <- coloc:::Var.data.cc(f = MAF, N = N1+N0, s = N1/(N0+N1)) # variance of estimated effect size

postprobs <- ppfunc(z = z0, V = var)

plot(postprobs, main = "Posterior probabilities of variants")
abline(v = iCV, col = 2)
```

The posterior probability plot shows the location of the causal variant (red line) - we see that the causal variant is not the variant with the largest posterior probability. This is sometimes seen in fine-mapping experiments due to linkage disequilibrium (correlation between SNPs).

The `credset` function reports the claimed coverage ^[Researchers commonly use the size of the credible set as an estimate of the coverage probabiltiy of the causal variant in the set. It is believed to be a slightly better indicator of coverage than the threshold (claimed coverage > threshold).] (the cumulative sum of the posterior probabilities of the variants in the set) and the number of variants in the credible set. If the `CV` parameter is supplied by the user (e.g. in simulation studies when the CV is known), then the output also includes a binary indicator of whether the causal variant was contained in the credible set. 

```{r warning=FALSE, error=FALSE}
thresh <- 0.9

credset(pp = postprobs,
      # CV = iCV,
       thr = thresh 
       ) 

cs <- credset(pp = postprobs,
              CV = iCV,
             thr = thresh 
       ) 

cs
```

The credible set obtained contains 114 (out of 200) variants and the claimed coverage is 0.9004596. In papers, we would expect to see author report that they have found a 90% credible set of 45 variants, which they are 90% confident contains the true causal variant. 

----
  
## Obtain corrected coverage estimate
  
----
  
Suppose we are suspicious of this coverage estimate and wish to find a more accurate coverage probability using this fancy new package. The only parameters we need to supply to the `corrcov` function are `z0` (marginal $Z$ scores), `f` (minor allele frequencies), `N0` (number of controls), `N1` (number of cases), `Sigma` (SNP correlation matrix) and the threshold of the credible set. The prior on the standard deviation of the effect size is an optional parameter, it's default value of $0.2$ is used here.

```{r cache = TRUE}
LD <- cor2(haps) # correlation between SNPs

corrected_cov_estimate <- corrcov(z0, f = MAF, N0, N1, Sigma = LD, thr = thresh)
```

```{r echo = FALSE, cache = TRUE}
data.frame("Corrected Coverage" = corrected_cov_estimate, "Claimed Coverage" = cs$claimed.cov, "Threshold" = thresh)
```

In this example, the 90% credible set may in actual fact have nearer to 96% coverage of the causal variant in the credible set. For the purpose of this vignette and to see the accuracy of this estimate, we simulate more credible sets from the same system and find what proportion of these contains the true causal variant. 

```{r}
z0.tmp <- simulated_z_score(N0=5000, # number of controls
                            N1=5000, # number of cases
                            snps=snps, # column names in freq
                            W=CV, # causal variants, subset of snps
                            gamma.W=log(g1), # log odds ratios
                            freq=freq, # reference haplotypes
                            nrep = 5000
)

pps <- ppfunc.mat(zstar = z0.tmp, V = var) # find pps
cs <- apply(pps, 1, credset, CV = iCV, thr = thresh) %>% data.table:::rbindlist() # form cred sets
true.cov.est <- mean(cs$covered) # average covered col
true.cov.est
```

The empirical coverage is found to be approximately 97.5%, so our estimate of ~96% is more accurate than the original estimate of 90% (or 90.09% is using claimed coverage).

```{r echo = FALSE, cache = TRUE}
data.frame("Empirical Coverage" = true.cov.est, "Corrected Coverage" = corrected_cov_estimate, "Claimed Coverage" = cs$claimed.cov, "Threshold" = thresh)[1,]
```

This vignette has shown readers how to use the `corrcoverage' package to obtain more accurate coverage estimates of the causal variant in the credible set.

----
  
## References

1. [simGWAS](https://github.com/chr1swallace/simGWAS): Mary D Fortune, Chris Wallace; simGWAS: a fast method for simulation of large scale case–control GWAS summary statistics, Bioinformatics, , bty898, https://doi.org/10.1093/bioinformatics/bty898 

```{r session_info, include=TRUE, echo=TRUE, results='markup'}
devtools::session_info()
```

