<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Provides corrected coverage value for the causal variant in a credible set • corrcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Provides corrected coverage value for the causal variant in a credible set">
<meta property="og:description" content="A package to obtain an accurate coverage estimate of the causal variant in a credible set obtained using the Bayesian approach to fine-mapping">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">corrcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/my-vignette.html">Corrected Coverage</a>
    </li>
    <li>
      <a href="articles/New-Credible-Set.html">New Credible Set</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="corrcoverage" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#corrcoverage" class="anchor"></a>corrcoverage</h1></div>
<p>The <code>corrcoverage</code> R package can be used to obtain an accurate coverage estimate of the causal variant in a credible set that has been acquired using the Bayesian approach for fine-mapping (<a href="https://www.ncbi.nlm.nih.gov/pubmed/23104008">Maller et al., 2012</a>, <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/gepi.20359">Wakefield, 2009</a>.).</p>
<p>Webpage: <a href="https://annahutch.github.io/corrcoverage/" class="uri">https://annahutch.github.io/corrcoverage/</a></p>
<hr>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of <code>corrcoverage</code> from <a href="https://github.com/">github</a> with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>) <span class="co"># if not already installed</span>
devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"annahutch/corrcoverage"</span>)</code></pre></div>
<hr>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>For examples and usage, please see the vignettes [<a href="http://annahutch.github.io/corrcoverage/articles/my-vignette.html" class="uri">http://annahutch.github.io/corrcoverage/articles/my-vignette.html</a>]</p>
<hr>
</div>
<div id="conversion-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#conversion-functions" class="anchor"></a>Conversion Functions</h2>
<p>The <code>corrcoverage</code> package includes useful functions for converting between p-values, <span class="math inline">\(Z\)</span>-scores, asymptotic Bayes factors and posterior probabilities of causality. The following table summarises these functions.</p>
<table class="table table">
<thead><tr class="header">
<th>Include null model? <a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>
</th>
<th>Input</th>
<th>Output</th>
<th>Function</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>YES</td>
<td>p-values</td>
<td>log(ABF)</td>
<td><code>approx.bf.p</code></td>
</tr>
<tr class="even">
<td>YES</td>
<td>p-values</td>
<td>Posterior Probabilities</td>
<td><code>pvals_pp</code></td>
</tr>
<tr class="odd">
<td>YES</td>
<td>Marginal <span class="math inline">\(Z\)</span>-scores</td>
<td>Posterior Probabilties</td>
<td><code>z0_pp</code></td>
</tr>
<tr class="even">
<td>NO</td>
<td>Marginal <span class="math inline">\(Z\)</span>-scores</td>
<td>Posterior Probabilties</td>
<td><code>ppfunc</code></td>
</tr>
<tr class="odd">
<td>NO</td>
<td>Marginal <span class="math inline">\(Z\)</span>-score matrix</td>
<td>Posterior Probabilties</td>
<td><code>z0func.mat</code></td>
</tr>
</tbody>
</table>
<p>Functions are also provided to simulate marginal <span class="math inline">\(Z\)</span>-scores from joint <span class="math inline">\(Z\)</span>-scores. The joint <span class="math inline">\(Z\)</span>-scores are all 0, except at the causal variant where it is the ‘true effect’, <span class="math inline">\(\mu\)</span>. <span class="math inline">\(\mu\)</span> can be estimated using <code>est_mu</code> function which required sample sizes, <span class="math inline">\(Z\)</span>-scores and minor allele frequencies.</p>
<p><code>z_sim</code> simulates marginal <span class="math inline">\(Z\)</span>-scores from joint <span class="math inline">\(Z\)</span>-scores, whilst <code>zj_pp</code> can be used to simulate posterior probability systems from a joint <span class="math inline">\(Z\)</span>-score vector. These functions work by first calculating <span class="math inline">\(E(Z_m)\)</span>, <span class="math display">\[E(Z_m)=Z_j \times \Sigma\]</span> where <span class="math inline">\(\Sigma\)</span> is the correlation matrix between SNPs.</p>
<p>We can then simulate more <span class="math inline">\(Z\)</span>-score systems from a multivariate normal disribution with mean <span class="math inline">\(E(Z_m)\)</span> and variance <span class="math inline">\(\Sigma\)</span>. This is a key step in our corrected coverage method.</p>
<hr>
<p>Our correction factor works by simulating many more credible sets from ‘the same system as the original’ and finding what proportion of these contain the true causal variant, whereby each variant is considered causal in turn and the predictions are normalised by that variant’s posterior probability of causality.</p>
<hr>
</div>
<div id="abstract" class="section level2">
<h2 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h2>
<p>The primary goal of Genome Wide Association Studies (GWAS) is to better understand the biology of disease. GWAS have been successful in identifying thousands of associations with common and complex diseases, but these associations refer to physical genomic regions rather than specific causal variants responsible for the disease. Consequently, known GWAS association signals are used in follow-up studies, which continue to untangle the relationship between genetic variation and disease. Accurately localising the specific “causal variants” driving the association signals identified by GWAS is difficult due to correlations between SNPs. Standard practice in fine-mapping experiments is therefore to report a credible set of variants which is believed to contain the causal variant with some “coverage probability”.</p>
<p>We evaluated coverage probabilities of credible sets obtained using the dominant method in the field and found that the claimed coverage probabilities were systematically biased. In low power studies the true coverage was below that claimed, suggesting that researchers should add more variants to the credible set in order to achieve the required coverage. In high power studies, the true coverage was higher than that claimed, potentially allowing for higher resolution through the removal of variants from the set, whilst still attaining the required coverage. The algorithm to create credible sets contains an ordering step, which aims to make the set as small as possible, but which is not accounted for when estimating coverage probabilities. We showed that it is this ordering step that induces coverage bias and have developed a method to estimate this bias using rapid simulations based on the observed SNP correlation structure. We provide R software for our method which provides the user with an accurate coverage estimate of the causal variant in the credible set. Obtaining accurate coverage estimates will allow researchers to adjust their credible set to achieve the true desired coverage - for example narrowing down a 95% credible set of 10 variants to just 5 variants.</p>
<p>While technical, our result - that standard coverage estimates of the causal variant in GWAS are inaccurate and can be improved - impacts standard practice in genetic association studies. Improvement in the resolution of the fine-mapping experiment will enable more efficient expenditure of resources in the follow-up process of annotating the variants in the credible set to determine the implicated genes and pathways, helping to untangle the complex relationship between genetic variants and disease.</p>
<hr>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>This refers to whether we include the null model of no genetic effect into the analyses. Note that in the standard Bayesian approach for fine-mapping, this is ignored such that the sum of the posterior probabilities for each variant is 1.<a href="#fnref1">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>0</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Anna Hutchinson <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
