<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>New Credible Set • corrcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="New Credible Set">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">corrcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/my-vignette.html">Corrected Coverage</a>
    </li>
    <li>
      <a href="../articles/New-Credible-Set.html">New Credible Set</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>New Credible Set</h1>
                        <h4 class="author">Anna Hutchinson</h4>
            
      
      
      <div class="hidden name"><code>New-Credible-Set.Rmd</code></div>

    </div>

    
    
<hr>
<p>This vignette will show users how the <code>corrcoverage</code> package can be used to obtain a new credible set of variants that contains the true causal variant with some specified desired coverage value whilst containing as few variants as possible.</p>
<hr>
<p>As before, let’s begin by simulating some GWAS data using the <code>simGWAS</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(simGWAS)

<span class="co"># Simulate reference haplotypes</span>
nsnps &lt;-<span class="st"> </span><span class="dv">200</span>
nhaps &lt;-<span class="st"> </span><span class="dv">1000</span>
lag &lt;-<span class="st"> </span><span class="dv">5</span> <span class="co"># genotypes are correlated between neighbouring variants</span>
maf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(nsnps<span class="op">+</span>lag,<span class="fl">0.05</span>,<span class="fl">0.5</span>) <span class="co"># common SNPs</span>
laghaps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="st">"cbind"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(maf, <span class="cf">function</span>(f) <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(nhaps,<span class="dv">1</span>,f)))
haps &lt;-<span class="st"> </span>laghaps[,<span class="dv">1</span><span class="op">:</span>nsnps]
<span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>lag) 
    haps &lt;-<span class="st"> </span>haps <span class="op">+</span><span class="st"> </span>laghaps[,(<span class="dv">1</span><span class="op">:</span>nsnps)<span class="op">+</span>j]
haps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(haps<span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(haps,<span class="dv">2</span>,max),nhaps,nsnps,<span class="dt">byrow=</span><span class="ot">TRUE</span>))
snps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(haps) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"s"</span>,<span class="dv">1</span><span class="op">:</span>nsnps)
freq &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(haps<span class="op">+</span><span class="dv">1</span>)
freq<span class="op">$</span>Probability &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(freq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(freq<span class="op">$</span>Probability)</code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAF &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(freq[,snps]<span class="op">-</span><span class="dv">1</span>) <span class="co"># minor allele frequencies</span>
CV &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(snps[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(haps)<span class="op">&gt;</span><span class="fl">0.1</span>)],<span class="dv">1</span>)
iCV &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">sub</a></span>(<span class="st">"s"</span>, <span class="st">""</span>, CV) <span class="co"># index of cv</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OR &lt;-<span class="st"> </span><span class="fl">1.1</span> <span class="co"># odds ratios</span>
z0 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/simGWAS/topics/simulated_z_score">simulated_z_score</a></span>(<span class="dt">N0=</span><span class="dv">5000</span>, <span class="co"># number of controls</span>
              <span class="dt">N1=</span><span class="dv">5000</span>, <span class="co"># number of cases</span>
              <span class="dt">snps=</span>snps, <span class="co"># column names in freq of SNPs for which Z scores should be generated</span>
              <span class="dt">W=</span>CV, <span class="co"># causal variants, subset of snps</span>
              <span class="dt">gamma.W=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR), <span class="co"># log odds ratios</span>
              <span class="dt">freq=</span>freq <span class="co"># reference haplotypes</span>
              )
N0 &lt;-<span class="st"> </span><span class="dv">5000</span> <span class="co"># number of controls</span>
N1 &lt;-<span class="st"> </span><span class="dv">5000</span> <span class="co"># number of cases</span>
thresh &lt;-<span class="st"> </span><span class="fl">0.9</span> <span class="co"># threshold</span>

var &lt;-<span class="st"> </span>coloc<span class="op">:::</span><span class="kw"><a href="../reference/Var.data.cc.html">Var.data.cc</a></span>(<span class="dt">f =</span> MAF, <span class="dt">N =</span> N1<span class="op">+</span>N0, <span class="dt">s =</span> N1<span class="op">/</span>(N0<span class="op">+</span>N1)) <span class="co"># variance of estimated effect size</span>

postprobs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ppfunc.html">ppfunc</a></span>(<span class="dt">z =</span> z0, <span class="dt">V =</span> var) <span class="co"># pp0</span>
LD &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cor2.html">cor2</a></span>(haps) <span class="co"># correlation matrix</span></code></pre></div>
<hr>
<p>We use the <code>est_mu</code> function to obtain an estimate of the true effect at the causal variant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">muhat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/est_mu.html">est_mu</a></span>(z0, MAF, N0, N1)</code></pre></div>
<hr>
<p>To speed things up, we simulate a list of matrices of posterior probabilities (one simulation per row), where each list element considers a different SNP as causal. For example, row 1 in the 2nd matrix/ list element is a posterior probability system simulated by considering SNP 2 as causal. The function <code>zj_pp</code> does this.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nsnps =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(postprobs)
temp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="dt">x =</span> muhat, <span class="dt">nrow =</span> nsnps, <span class="dt">ncol =</span> nsnps)
zj &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(c, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(temp, <span class="dv">1</span>, list))  <span class="co"># nsnp zj vectors for each snp considered causal</span>

<span class="co"># simulate pp systems</span>
pps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mapply">mapply</a></span>(zj_pp, zj, <span class="dt">V =</span> var , <span class="dt">MoreArgs =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">nrep =</span> <span class="dv">1000</span>, <span class="dt">Sigma =</span> LD), <span class="dt">SIMPLIFY =</span> <span class="ot">FALSE</span>)</code></pre></div>
<hr>
<p>The <code>quick_corrcov_cs</code> function can now be used to quickly find the credible set and the corrected coverage estimate of this credible set using a threshold defined by the user.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/quick_corrcov_cs.html">quick_corrcov_cs</a></span>(<span class="fl">0.8</span>, <span class="dt">simulated.pps =</span> pps, <span class="dt">pp =</span> postprobs)</code></pre></div>
<pre><code>## $credset
##   [1] "s43"  "s49"  "s33"  "s25"  "s55"  "s26"  "s136" "s45"  "s83"  "s197"
##  [11] "s160" "s39"  "s23"  "s195" "s73"  "s20"  "s19"  "s184" "s18"  "s161"
##  [21] "s8"   "s164" "s34"  "s196" "s92"  "s44"  "s91"  "s200" "s17"  "s98" 
##  [31] "s1"   "s5"   "s70"  "s114" "s158" "s183" "s30"  "s16"  "s170" "s42" 
##  [41] "s129" "s31"  "s169" "s157" "s186" "s38"  "s52"  "s78"  "s165" "s167"
##  [51] "s85"  "s99"  "s21"  "s139" "s96"  "s105" "s71"  "s162" "s32"  "s66" 
##  [61] "s187" "s194" "s53"  "s54"  "s4"   "s168" "s109" "s198" "s179" "s48" 
##  [71] "s130" "s175" "s159" "s12"  "s51"  "s142" "s40"  "s166" "s131" "s144"
##  [81] "s119" "s59"  "s57"  "s190" "s28"  "s116" "s47"  "s74"  "s151" "s35" 
##  [91] "s156" "s135" "s61"  "s138" "s134" "s3"   "s118" "s81"  "s171" "s140"
## [101] "s68"  "s199" "s115" "s77"  "s141" "s108" "s133" "s15"  "s180" "s41" 
## [111] "s90"  "s149"
## 
## $corr.cov
## [1] 0.8794729
## 
## $thr
## [1] 0.8
## 
## $size
##      s149 
## 0.8022306</code></pre>
<hr>
<p>The <code>XXX</code> function can be used to find the smallest credible set that has corrected coverage as close to the user specified desired coverage as possible (specify accuracy…)</p>
<p>STILL WORKING ON THIS</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
