<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>From GWAS Summary Statistics to Credible Sets • corrcoverage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="From GWAS Summary Statistics to Credible Sets">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">corrcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_gwas_to_cs.html">GWAS to Credible Set</a>
    </li>
    <li>
      <a href="../articles/Corrected-Coverage.html">Corrected Coverage</a>
    </li>
    <li>
      <a href="../articles/New-Credible-Set.html">New Credible Set</a>
    </li>
    <li>
      <a href="../articles/Useful-Info.html">Additional Useful Functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/annahutch/corrcoverage">
    <span class=" github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>From GWAS Summary Statistics to Credible Sets</h1>
                        <h4 class="author">Anna Hutchinson</h4>
            
      
      
      <div class="hidden name"><code>a01_gwas_to_cs.Rmd</code></div>

    </div>

    
    
<p><a href="https://www.nature.com/articles/ng.2435">Maller et al. (2012)</a> proposed a method to find credible sets of putative causal variants, assuming that there is only one causal variant (CV) per associated region and that this is typed in the study. This method can be used in instances where only single variant summary statistics are available thanks to <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/gepi.20359">Wakefield’s approximation of Bayes factors</a>.</p>
<p>Let <span class="math inline">\(\beta_i\)</span>, for <span class="math inline">\(i=1,...,k\)</span> SNPs in a genomic region, be the regression coefficient from a single-SNP logistic regression model, quantifying the evidence of an association between SNP <span class="math inline">\(i\)</span> and the disease. Assuming that there is only one CV per region and that this is typed in the study, then if SNP <span class="math inline">\(i\)</span> is causal, <span class="math inline">\(\beta_i\neq 0\)</span> and <span class="math inline">\(\beta_j\)</span> (for <span class="math inline">\(j\neq i\)</span>) is non-zero only through LD between SNPs <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>. Note that no parametric assumptions are required for <span class="math inline">\(\beta_i\)</span> yet, so we write that it is sampled from some distribution, <span class="math inline">\(\beta_i \sim []\)</span>. We can then write down the likelihood, <span class="math display">\[\begin{equation}
\begin{split}
        P(D|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) &amp; = P(D_i
|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) \times P(D_{-i}|D_i,\text{ }\beta_i\sim\text{[ ]},\text{ }i\text{ causal})\\
    &amp; = P(D_i
|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) \times P(D_{-i}|D_i,\text{ }i\text{ causal})\,,
\end{split}
    \end{equation}\]</span></p>
<p>since <span class="math inline">\(D_{-i}\)</span> is independent of <span class="math inline">\(\beta_i\)</span> given <span class="math inline">\(D_i\)</span>. Here, <span class="math inline">\(D\)</span> is the genotype data (0, 1 or 2 counts of the minor allele) for the entire genomic region and <span class="math inline">\(i\)</span> is a SNP in the region, such that <span class="math inline">\(D_i\)</span> and <span class="math inline">\(D_{-i}\)</span> are the genotype data at SNP <span class="math inline">\(i\)</span> and at the remaining SNPs in the genomic region, respectively.</p>
<p>We can now place some parametric assumptions on SNP <span class="math inline">\(i\)</span>’s true effect on disease. This is typically quantified as log odds ratio (OR), and is assumed to be sampled from a Gaussian distribution, <span class="math inline">\(\beta_i\sim N(0,W)\)</span>, where <span class="math inline">\(W\)</span> is chosen to reflect the researcher’s prior belief on the variability of the true OR. We chose to set to <span class="math inline">\(W=0.2\)</span> in our method, reflecting a belief that 95% of ORs range from <span class="math inline">\(exp(-1.96\times 0.2)=0.68\)</span> to <span class="math inline">\(exp(1.96\times 0.2)=1.48\)</span>.</p>
<p>The posterior probabilities of causality (PP) for each SNP <span class="math inline">\(i\)</span> in an associated genomic region with <span class="math inline">\(k\)</span> SNPs can be calculated where, <span class="math display">\[\begin{equation}
        PP_i=P(\beta_i \sim N(0,W),\text{ }i \text{ causal}|D)\,, \quad i \in \{1,...,k\}.
\end{equation}\]</span></p>
<p>Under the assumption that each SNP is  equally likely to be causal, then <span class="math display">\[\begin{equation}
P(\beta_i \sim N(0,W),\text{ }i\text{ causal})=\dfrac{1}{k}\,, \quad i \in \{1,...,k\}
\end{equation}\]</span> and Bayes theorem can be used to write <span class="math display">\[\begin{equation}
\begin{aligned}
    PP_i=P(\beta_i \sim N(0,W),\text{ }i \text{ causal}|D)\propto P(D|\beta_i\sim N(0,W),\text{ }i\text{ causal}).
\end{aligned}
\end{equation}\]</span></p>
<p>Dividing through by the probability of the genotype data given the null model of no genetic effect, <span class="math inline">\(H_0\)</span>, yields a likelihood ratio, <span class="math display">\[\begin{equation}
    PP_i\propto \dfrac{P(D|\beta_i \sim N(0,W),\text{ }i \text{ causal)}}{P(D|H_0)},
\end{equation}\]</span></p>
<p>from which Equation (1) can be used to derive, <span class="math display">\[\begin{equation}
    PP_i\propto \frac{P(D_i|\beta_i \sim N(0,W),\text{ }i \text{ causal})}{P(D_i|H_0)}= BF_i\,,
\end{equation}\]</span> where <span class="math inline">\(BF_i\)</span> is the Bayes factor for SNP <span class="math inline">\(i\)</span>, measuring the ratio of the probabilities of the data at SNP <span class="math inline">\(i\)</span> given the alternative (SNP <span class="math inline">\(i\)</span> is causal) and the null (no genetic effect) models.</p>
<p>In genetic association studies where sample sizes are usually large, these BFs can be approximated using Wakefield’s asymptotic Bayes factors (ABFs). Given that <span class="math inline">\(\hat\beta_i\sim N(\beta_i,V_i)\)</span> and  <span class="math inline">\(\beta_i\sim N(0,W)\)</span>,</p>
<p><span class="math display">\[\begin{equation}
    ABF_i=\sqrt{\frac{V_i}{V_i+W}}exp\left(\frac{Z_i^2}{2}\frac{W}{(V_i+W)}\right)\,,
\end{equation}\]</span> where <span class="math inline">\(Z_i^2=\dfrac{\hat\beta_i^2}{V_i}\)</span> is the squared marginal <span class="math inline">\(Z\)</span> score for SNP <span class="math inline">\(i\)</span>.</p>
<p>In Bayesian fine-mapping, PPs are calculated for all SNPs in the genomic region and the variants are sorted into descending order of their PP. The PPs are then cumulatively summed until some threshold, <span class="math inline">\(\alpha\)</span>, is exceeded. The variants required to exceed this threshold form the <span class="math inline">\(\alpha\)</span>-level credible set.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
