<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>New Credible Set: <code>corrected_cs</code> • corrcoverage</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="New Credible Set: &lt;code&gt;corrected_cs&lt;/code&gt;">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">corrcoverage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/New-Credible-Set.html">New Credible Set: `corrected_cs`</a>
    </li>
    <li>
      <a href="../articles/Useful-Info.html">Useful Info</a>
    </li>
    <li>
      <a href="../articles/my-vignette.html">Correcting the Coverage of Credible Sets: `corrcov`</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>New Credible Set: <code>corrected_cs</code>
</h1>
                        <h4 class="author">Anna Hutchinson</h4>
            
      
      
      <div class="hidden name"><code>New-Credible-Set.Rmd</code></div>

    </div>

    
    
<p>This vignette will show users how the <code>corrcoverage</code> package can be used to obtain a new credible set of variants that contains the true causal variant with some specified desired coverage value whilst containing as few variants as possible.</p>
<hr>
<div id="simulate-gwas-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simulate-gwas-data" class="anchor"></a>Simulate GWAS data</h2>
<p>As in the <a href="https://annahutch.github.io/corrcoverage/articles/my-vignette.html">corrected coverage vignette</a>, let’s begin by simulating some GWAS data using the <code>simGWAS</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(simGWAS)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#  Simulate reference haplotypes</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">nsnps &lt;-<span class="st"> </span><span class="dv">200</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">nhaps &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">lag &lt;-<span class="st"> </span><span class="dv">30</span>  <span class="co"># genotypes are correlated between neighbouring variants</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">maf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(nsnps <span class="op">+</span><span class="st"> </span>lag, <span class="fl">0.05</span>, <span class="fl">0.5</span>)  <span class="co"># common SNPs</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">laghaps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="st">"cbind"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(maf, <span class="cf">function</span>(f) <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(nhaps, <span class="dv">1</span>, f)))</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">haps &lt;-<span class="st"> </span>laghaps[, <span class="dv">1</span><span class="op">:</span>nsnps]</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>lag) haps &lt;-<span class="st"> </span>haps <span class="op">+</span><span class="st"> </span>laghaps[, (<span class="dv">1</span><span class="op">:</span>nsnps) <span class="op">+</span><span class="st"> </span>j]</a>
<a class="sourceLine" id="cb1-11" data-line-number="11">haps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(haps<span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(haps, <span class="dv">2</span>, max), nhaps, nsnps, <span class="dt">byrow =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-12" data-line-number="12">snps &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(haps) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"s"</span>, <span class="dv">1</span><span class="op">:</span>nsnps)</a>
<a class="sourceLine" id="cb1-13" data-line-number="13">freq &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(haps <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-14" data-line-number="14">freq<span class="op">$</span>Probability &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(freq)</a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(freq<span class="op">$</span>Probability)</a></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">MAF &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(freq[, snps] <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)  <span class="co"># minor allele frequencies</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">CV &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(snps[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(haps) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.1</span>)], <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">iCV &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">sub</a></span>(<span class="st">"s"</span>, <span class="st">""</span>, CV)  <span class="co"># index of cv</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">LD &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cor2.html">cor2</a></span>(haps) <span class="co"># correlation between SNPs</span></a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">OR &lt;-<span class="st"> </span><span class="fl">1.1</span> <span class="co"># odds ratios</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">N0 &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="co"># number of controls</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">N1 &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="co"># number of cases</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  </a>
<a class="sourceLine" id="cb4-5" data-line-number="5">z0 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/simGWAS/topics/simulated_z_score">simulated_z_score</a></span>(<span class="dt">N0 =</span> N0, <span class="co"># number of controls</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">                        <span class="dt">N1 =</span> N1, <span class="co"># number of cases</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">                        <span class="dt">snps =</span> snps, <span class="co"># column names in freq of SNPs for which Z scores should be generated</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8">                        <span class="dt">W =</span> CV, <span class="co"># causal variants, subset of snps</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">                        <span class="dt">gamma.W =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR), <span class="co"># log odds ratios</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">                        <span class="dt">freq =</span> freq) <span class="co"># reference haplotypes</span></a></code></pre></div>
<p>To calculate <span class="math inline">\(V\)</span>, the prior variance for the estimated effect size, we use <code>Var.data.cc</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">varbeta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Var.data.cc.html">Var.data.cc</a></span>(<span class="dt">f =</span> MAF, <span class="dt">N =</span> N1<span class="op">+</span>N0, <span class="dt">s =</span> N1<span class="op">/</span>(N0<span class="op">+</span>N1)) <span class="co"># variance of estimated effect size</span></a></code></pre></div>
<p>We can then use the <code>ppfunc</code> function to calculate the posterior probabilities of causality for each variant.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">postprobs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ppfunc.html">ppfunc</a></span>(<span class="dt">z =</span> z0, <span class="dt">V =</span> varbeta) </a></code></pre></div>
<p>We use the <code>est_mu</code> function to obtain an estimate of the true effect at the causal variant.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">muhat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/est_mu.html">est_mu</a></span>(z0, MAF, N0, N1)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">muhat</a></code></pre></div>
<pre><code>## [1] 4.970273</code></pre>
<hr>
</div>
<div id="corrected-coverage-estimate" class="section level2">
<h2 class="hasAnchor">
<a href="#corrected-coverage-estimate" class="anchor"></a>Corrected Coverage Estimate</h2>
<p>The <code>corrected_cov</code> function is used to find the corrected coverage of a credible set with specified threshold, say 0.9.</p>
<p>Note that this function is similar to using <code>corrcov</code> as explained in the <a href="https://annahutch.github.io/corrcoverage/articles/my-vignette.html">“Corrected Coverage” vignette</a>; which would require <span class="math inline">\(Z\)</span>-scores, minor allele frequencies and sample sizes. Here, we have already calculated some of the intermediaries calculated in the <code>corrcov</code> function (muhat, varbeta etc.) so we can use <code>corrected_cov</code> instead.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">thr =<span class="st"> </span><span class="fl">0.9</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">corrcov &lt;-<span class="st"> </span><span class="kw"><a href="../reference/corrected_cov.html">corrected_cov</a></span>(<span class="dt">pp0 =</span> postprobs, <span class="dt">mu =</span> muhat, <span class="dt">V =</span> varbeta, <span class="dt">Sigma =</span> LD, <span class="dt">thr =</span> thr, <span class="dt">nrep =</span> <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">cs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/credset.html">credset</a></span>(<span class="dt">pp =</span> postprobs, <span class="dt">thr =</span> thr)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">claimed.cov =</span> cs<span class="op">$</span>claimed.cov, <span class="dt">corr.cov =</span>  corrcov, <span class="dt">nvar =</span> cs<span class="op">$</span>nvar)</a></code></pre></div>
<pre><code>##   claimed.cov  corr.cov nvar
## 1   0.9307395 0.9672825    9</code></pre>
<p>Using the Bayesian approach for statistical fine-mapping we obtain a 90% credible set consisting of 9 variants. The claimed coverage of this credible set is ~0.93, yet the corrected coverage estimate is ~0.97, suggesting that we can afford to be ‘more confident’ that we have captured the causal variant in our credible set.</p>
</div>
<div id="empirical-coverage-estimate" class="section level2">
<h2 class="hasAnchor">
<a href="#empirical-coverage-estimate" class="anchor"></a>Empirical Coverage Estimate</h2>
<p>Again, for the purpose of this vignette we can investigate how accurate this estimate is by simulating many credible sets from the same system, and finding the proportion of these that contain the true causal variant.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">z0.tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/simGWAS/topics/simulated_z_score">simulated_z_score</a></span>(<span class="dt">N0 =</span> N0, <span class="co"># number of controls</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">                            <span class="dt">N1 =</span> N1, <span class="co"># number of cases</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                            <span class="dt">snps =</span> snps, <span class="co"># column names in freq of SNPs for which Z scores should be generated</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4">                            <span class="dt">W =</span> CV, <span class="co"># causal variants, subset of snps</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5">                            <span class="dt">gamma.W =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR), <span class="co"># log odds ratios</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6">                            <span class="dt">freq =</span> freq, <span class="co"># reference haplotypes</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7">                            <span class="dt">nrep =</span> <span class="dv">5000</span>) </a>
<a class="sourceLine" id="cb11-8" data-line-number="8"></a>
<a class="sourceLine" id="cb11-9" data-line-number="9">pps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ppfunc.mat.html">ppfunc.mat</a></span>(<span class="dt">zstar =</span> z0.tmp, <span class="dt">V =</span> varbeta)  <span class="co"># find pps from a matrix of simulations</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10">cs.cov &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(pps, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw"><a href="../reference/credset.html">credset</a></span>(x, <span class="dt">CV =</span> iCV, <span class="dt">thr =</span> thr)<span class="op">$</span>cov)</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">true.cov.est &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(cs.cov)</a>
<a class="sourceLine" id="cb11-12" data-line-number="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">claimed.cov =</span> cs<span class="op">$</span>claimed.cov, <span class="dt">corr.cov =</span>  corrcov, <span class="dt">true.cov =</span> true.cov.est, <span class="dt">nvar =</span> cs<span class="op">$</span>nvar)</a></code></pre></div>
<pre><code>##   claimed.cov  corr.cov true.cov nvar
## 1   0.9307395 0.9672825   0.9696    9</code></pre>
<p>We find that our corrected coverage value is very close to the empirical coverage of the credible set.</p>
<hr>
</div>
<div id="obtain-a-new-credible-set" class="section level2">
<h2 class="hasAnchor">
<a href="#obtain-a-new-credible-set" class="anchor"></a>Obtain a New Credible Set</h2>
<p>Our results suggest that we may be able to remove some variants from the credible set, whilst still achieving the desired coverage of 90%.</p>
<p>The <code>corrected_cs</code> function uses GWAS summary statistics and some user-defined parameters to find the smallest credible set such that the coverage estimate is within some accuracy of the desired coverage.</p>
<p>The function requires the <span class="math inline">\(Z\)</span>-scores (<code>z</code>), minor allele frequencies (<code>f</code>), control and case sample sizes (<code>N0</code> and <code>N1</code> respectively), LD matrix (<code>Sigma</code>), some lower threshold value (<code>lower</code>), some upper threshold value (<code>upper</code>) and the desired coverage (<code>desired.cov</code>).</p>
<p>The function uses the <a href="https://en.wikipedia.org/wiki/Root-finding_algorithm#Bisection_method">bisection root finding method</a> to converge to the smallest threshold such that the corrected coverage is larger than the desired coverage. The accuracy parameter is optional (default value set to 0.005) and controls how much greater than the desired coverage the user is happy for the corrected coverage value to lie.</p>
<p>The function reports the threshold values tested and their corresponding corrected coverage value. The maximum number of iterations for the bisecting root finding algorithm is an optional parameter, with default value 20. The functions stops when either the number of iterations reaches the maximum, or the corrected coverage is within some accuracy of the desired coverage.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/corrected_cs.html">corrected_cs</a></span>(<span class="dt">z =</span> z0, <span class="dt">f =</span> MAF, N0, N1, <span class="dt">Sigma =</span> LD, <span class="dt">lower =</span> <span class="fl">0.5</span>, <span class="dt">upper =</span> <span class="dv">1</span>, <span class="dt">desired.cov =</span> <span class="fl">0.9</span>)</a></code></pre></div>
<pre><code>## [1] "thr:  0.75 , cov:  0.917720376269878"
## [1] "thr:  0.625 , cov:  0.878893777630186"
## [1] "thr:  0.6875 , cov:  0.896800428830677"
## [1] "thr:  0.71875 , cov:  0.907035299005275"
## [1] "thr:  0.703125 , cov:  0.901704434955293"</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">res</a></code></pre></div>
<pre><code>## $credset
## [1] "s54" "s58" "s57" "s67"
## 
## $req.thr
## [1] 0.703125
## 
## $corr.cov
## [1] 0.9017044
## 
## $size
## [1] 0.7343958</code></pre>
<p>In this example we see that a much smaller threshold value is required to obtain a credible set with 90% corrected coverage of the causal variant, containing only 4 variants. In the standard Bayesian approach, a threshold of 90% is chosen which we have seen leads to over-coverage.</p>
<p>Finally, we can compare this coverage estimate of the corrected credible set to an empirical estimate of the true coverage.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">new.cs.sims &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(pps, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw"><a href="../reference/credset.html">credset</a></span>(x, <span class="dt">CV =</span> iCV, <span class="dt">thr =</span> res<span class="op">$</span>req.thr)<span class="op">$</span>cov)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">true.cov.est2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(new.cs.sims)</a></code></pre></div>
<p>Original 90% credible set:</p>
<pre><code>##  claimed.cov corr.cov true.cov nvar
##        0.931    0.967     0.97    9</code></pre>
<p>New 90% credible set:</p>
<pre><code>##  claimed.cov corr.cov true.cov nvar
##        0.734    0.902    0.897    4</code></pre>
<hr>
<p>This vignette has shown how the <code>corrcoverage</code> R package can be used to improve the resolution of a credible set from Bayesian genetic fine-mapping, without the use of any additional data.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulate-gwas-data">Simulate GWAS data</a></li>
      <li><a href="#corrected-coverage-estimate">Corrected Coverage Estimate</a></li>
      <li><a href="#empirical-coverage-estimate">Empirical Coverage Estimate</a></li>
      <li><a href="#obtain-a-new-credible-set">Obtain a New Credible Set</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Hutchinson, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
